# Importación de datos



In [1]:
import pandas as pd

def cargarDatos() -> tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    """
    Carga los datasets de las tiendas desde archivos locales o URLs de respaldo.

    Return:
    - Tupla con cuatro DataFrames en el orden: tienda1, tienda2, tienda3, tienda4.
    """
    try:
        # Intenta cargar localmente primero.
        print("Función iniciada.")
        print("Cargando archivos localmente.")
        folder_name = "base-de-datos-challenge1-latam/"
        tienda_1 = pd.read_csv(folder_name + "tienda_1.csv")
        tienda_2 = pd.read_csv(folder_name + "tienda_2.csv")
        tienda_3 = pd.read_csv(folder_name + "tienda_3.csv")
        tienda_4 = pd.read_csv(folder_name + "tienda_4.csv")

    except FileNotFoundError:
        # Fallback a URLs (para Colab o si faltan archivos locales).
        print("No fue posible cargar uno o varios archivos locales.")
        print("Cargando archivos remotamente.")
        try: 
            base_url = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/"
            tienda_1 = pd.read_csv(base_url + "tienda_1%20.csv")
            tienda_2 = pd.read_csv(base_url + "tienda_2.csv")
            tienda_3 = pd.read_csv(base_url + "tienda_3.csv")
            tienda_4 = pd.read_csv(base_url + "tienda_4.csv")
            print("Archivos cargados remotamente.")
            return tienda_1, tienda_2, tienda_3, tienda_4
        except Exception as e:
            print("No fue posible cargar los datos. Verifique la excepción.")
            print(f'Excepción: {e}')
            
    except Exception as e:
        print("No fue posible cargar los datos. Verifique la excepción.")
        print(f'Excepción: {e}')

    else:
        print("Archivos cargados localmente.")    
        return tienda_1, tienda_2, tienda_3, tienda_4
    
    finally:
        print("Función terminada")

In [2]:
tienda1, tienda2, tienda3, tienda4 = cargarDatos()
tienda1.head()

Función iniciada.
Cargando archivos localmente.
Archivos cargados localmente.
Función terminada


,Producto,Categoría del Producto,Precio,Costo de envío,Fecha de Compra,Vendedor,Lugar de Compra,Calificación,Método de pago,Cantidad de cuotas,lat,lon
0,Asistente virtual,Electrónicos,164300.0,6900.0,16/01/2021,Pedro Gomez,Bogotá,4,Tarjeta de crédito,8,4.60971,-74.08175
1,Mesa de comedor,Muebles,192300.0,8400.0,18/05/2022,Beatriz Morales,Medellín,1,Tarjeta de crédito,4,6.25184,-75.56359
2,Juego de mesa,Juguetes,209600.0,15900.0,15/03/2021,Juan Fernandez,Cartagena,1,Tarjeta de crédito,1,10.39972,-75.51444
3,Microondas,Electrodomésticos,757500.0,41000.0,03/05/2022,Juan Fernandez,Cali,4,Nequi,1,3.43722,-76.52250
4,Silla de oficina,Muebles,335200.0,20200.0,07/11/2020,Maria Alfonso,Medellín,5,Nequi,1,6.25184,-75.56359


## Conozca el conjunto de datos

El objetivo es comparar las 4 tiendas, así que debemos asegurarnos que los DataFrames (DFs) contienen el mismo número de columnas, en el mismo orden y con el mismo tipo de datos.
Como referencia inicial usaremos la Tienda 1.

In [3]:
tienda1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2359 entries, 0 to 2358
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Producto                2359 non-null   object 
 1   Categoría del Producto  2359 non-null   object 
 2   Precio                  2359 non-null   float64
 3   Costo de envío          2359 non-null   float64
 4   Fecha de Compra         2359 non-null   object 
 5   Vendedor                2359 non-null   object 
 6   Lugar de Compra         2359 non-null   object 
 7   Calificación            2359 non-null   int64  
 8   Método de pago          2359 non-null   object 
 9   Cantidad de cuotas      2359 non-null   int64  
 10  lat                     2359 non-null   float64
 11  lon                     2359 non-null   float64
dtypes: float64(4), int64(2), object(6)
memory usage: 221.3+ KB


Ahora sabemos que en teoría cada DF debe tener 12 columnas:
- 4 de tipo float
- 2 de tipo int64
- 6 de tipo obj

Comprobamos que el index (nombre de columna) y orden de las columnas sea el mismo para los 4 DFs mediante el método equals():

In [4]:
tienda1.columns

Index(['Producto', 'Categoría del Producto', 'Precio', 'Costo de envío',
       'Fecha de Compra', 'Vendedor', 'Lugar de Compra', 'Calificación',
       'Método de pago', 'Cantidad de cuotas', 'lat', 'lon'],
      dtype='object')

In [5]:
type(tienda1.columns)

pandas.core.indexes.base.Index

In [6]:
tienda1.columns.equals(tienda2.columns) and \
tienda1.columns.equals(tienda3.columns) and \
tienda1.columns.equals(tienda4.columns)

True

Comprobamos que el tipo de dato de las columnas sea el mismo para los 4 DFs:

In [7]:
tienda1.dtypes

Producto                   object
Categoría del Producto     object
Precio                    float64
Costo de envío            float64
Fecha de Compra            object
Vendedor                   object
Lugar de Compra            object
Calificación                int64
Método de pago             object
Cantidad de cuotas          int64
lat                       float64
lon                       float64
dtype: object

In [8]:
type(tienda1.dtypes)

pandas.core.series.Series

In [9]:
tienda1.dtypes.equals(tienda2.dtypes) \
and tienda1.dtypes.equals(tienda3.dtypes) \
and tienda1.dtypes.equals(tienda4.dtypes)

True

En conclusión, los 4 DFs cuentan con la misma estructura, es decir, mismas columnas, mismo order y mismo tipo de datos.

# Concatenación de DataFrames

Para hacer el análisis más práctico y ahorrarnos una líneas de código, vamos a crear un DF que contenga los datos de todas las tiendas.

Para distinguir un DF de otro, y poder agruparlos en el futuro, vamos a crear una columna que indique a que tienda pertenece y después vamos a concatenar todos los DFs en uno sólo.

In [10]:
tienda1["Tienda"] = 1
tienda2["Tienda"] = 2
tienda3["Tienda"] = 3
tienda4["Tienda"] = 4

In [11]:
tiendas = pd.concat([tienda1, tienda2, tienda3, tienda4])
tiendas

,Producto,Categoría del Producto,Precio,Costo de envío,Fecha de Compra,Vendedor,Lugar de Compra,Calificación,Método de pago,Cantidad de cuotas,lat,lon,Tienda
0,Asistente virtual,Electrónicos,164300.0,6900.0,16/01/2021,Pedro Gomez,Bogotá,4,Tarjeta de crédito,8,4.60971,-74.08175,1
1,Mesa de comedor,Muebles,192300.0,8400.0,18/05/2022,Beatriz Morales,Medellín,1,Tarjeta de crédito,4,6.25184,-75.56359,1
2,Juego de mesa,Juguetes,209600.0,15900.0,15/03/2021,Juan Fernandez,Cartagena,1,Tarjeta de crédito,1,10.39972,-75.51444,1
3,Microondas,Electrodomésticos,757500.0,41000.0,03/05/2022,Juan Fernandez,Cali,4,Nequi,1,3.43722,-76.52250,1
4,Silla de oficina,Muebles,335200.0,20200.0,07/11/2020,Maria Alfonso,Medellín,5,Nequi,1,6.25184,-75.56359,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2353,Mesa de comedor,Muebles,227700.0,12600.0,12/02/2020,Camila Rivera,Medellín,5,Tarjeta de crédito,1,6.25184,-75.56359,4
2354,Silla de oficina,Muebles,342000.0,18600.0,17/05/2022,Felipe Santos,Bucaramanga,5,Tarjeta de crédito,3,7.12539,-73.11980,4
2355,Vaso térmico,Deportes y diversión,36700.0,2400.0,03/05/2022,Camila Rivera,Bogotá,5,Tarjeta de crédito,1,4.60971,-74.08175,4
2356,Set de ollas,Artículos para el hogar,151800.0,8300.0,15/06/2020,Blanca Ramirez,Cali,5,Tarjeta de crédito,1,3.43722,-76.52250,4


Como curiosidad, es posible crear un índice personalizado.

In [12]:
# tienda5 = tienda1.copy()
# tienda5["Tienda"] = 5
# tienda5.index = [f'1-{i}' for i in range(len(tienda5))]
# tienda5

# Tratamiento de valores nulos

Es importante comprobar si nuestro DF consolidado tiene valores nulos, en dado caso, se eliminarán dichos registros.

Notar el mensaje de error.

In [13]:
tiendas.groupby("Tienda").isnull().sum()

AttributeError: 'DataFrameGroupBy' object has no attribute 'isnull'

La mejor opción es aplicar apply() para lograr agrupar todos los datos en base al número de "Tienda", y a su vez, dentro de cada tienda agrupar el conteo de los valores nulos por columna.

In [ ]:
tiendas.groupby("Tienda").apply(lambda x: x.isnull().sum(), include_groups=False)

,Producto,Categoría del Producto,Precio,Costo de envío,Fecha de Compra,Vendedor,Lugar de Compra,Calificación,Método de pago,Cantidad de cuotas,lat,lon
Tienda,,,,,,,,,,,,
1,0,0,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,0,0


En conclusión, no hay valores nulos, no es necesario ninguna modificación

# Tratamiento de valores inconsistentes

A mi consideración, para saber si un valor es o no inconsistente, sería necesario consultar con un experto en el negocio (Señor Juan) que nos pueda orientar, especialmente si es un dato tipo obj.\
\
Y si es un valor númerico sería bueno saber si hay un domain o límite máximo y mínimo para ese valor.

Por el momento me guiaré con el sentido común, por ejemplo, si "Lugar de Compra" no fuese un lugar real, o si "Precio" tuviese un valor negativo, dichos registros serían eliminados o corregidos.

Para mayor facilidad de lectura establecemos un formato que use comas como separadores para datos tipo float.

In [ ]:
pd.options.display.float_format = '{:,.2f}'.format

In [ ]:
tiendas.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9435 entries, 0 to 2357
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Producto                9435 non-null   object 
 1   Categoría del Producto  9435 non-null   object 
 2   Precio                  9435 non-null   float64
 3   Costo de envío          9435 non-null   float64
 4   Fecha de Compra         9435 non-null   object 
 5   Vendedor                9435 non-null   object 
 6   Lugar de Compra         9435 non-null   object 
 7   Calificación            9435 non-null   int64  
 8   Método de pago          9435 non-null   object 
 9   Cantidad de cuotas      9435 non-null   int64  
 10  lat                     9435 non-null   float64
 11  lon                     9435 non-null   float64
 12  Tienda                  9435 non-null   int64  
dtypes: float64(4), int64(3), object(6)
memory usage: 1.0+ MB


## Columnas cuantitativas

## 'Precio', 'Costo de envío', 'Calificación', 'Cantidad de cuotas', 'lat', 'lon', 'Tienda'

Para las columnas cuantitativas bastara con usar el método describe():

In [ ]:
tiendas.describe()

,Precio,Costo de envío,Calificación,Cantidad de cuotas,lat,lon,Tienda
count,"9,435.00","9,435.00","9,435.00","9,435.00","9,435.00","9,435.00","9,435.00"
mean,"466,732.29","24,875.15",4.01,2.93,5.43,-74.76,2.50
std,"590,320.68","31,480.04",1.39,2.80,2.36,1.21,1.12
min,"7,600.00",0.00,1.00,1.00,-4.22,-77.28,1.00
25%,"55,200.00","3,000.00",3.00,1.00,4.61,-75.56,1.50
50%,"221,500.00","11,700.00",5.00,1.00,4.61,-74.20,2.00
75%,"663,300.00","34,900.00",5.00,4.00,6.25,-74.08,3.00
max,"2,977,000.00","162,500.00",5.00,24.00,11.54,-67.92,4.00


No hay algún outlier o valor atípico apreciable.\
Cabe resaltar que: -90 <= latitud <= 90.\
Cabe resaltar que: -180 <= longitud <= 180.

## Columnas cualitativas

Para la mayoría de columnas cualitativas, simplemente agruparemos sus valores y los contaremos.

### 'Producto'

In [ ]:
tiendas.groupby("Producto")["Producto"].value_counts()

Producto
Ajedrez de madera              180
Armario                        197
Asistente virtual              167
Auriculares                    174
Auriculares con micrófono      158
Balón de baloncesto            197
Balón de voleibol              184
Batería                        203
Bicicleta                      196
Bloques de construcción        199
Cama box                       199
Cama king                      201
Carrito de control remoto      206
Celular ABXY                   157
Ciencia de datos con Python    166
Cubertería                     189
Cubo mágico 8x8                180
Cuerda para saltar             186
Dashboards con Power BI        181
Dinosaurio Rex                 175
Estufa                         187
Guitarra acústica              188
Guitarra eléctrica             165
Impresora                      188
Iniciando en programación      195
Iphone 15                      198
Juego de mesa                  175
Kit de bancas                  188
Lavadora de

### 'Categoría del Producto'

In [ ]:
tiendas.groupby("Categoría del Producto")["Categoría del Producto"].value_counts()

Categoría del Producto
Artículos para el hogar     730
Deportes y diversión       1113
Electrodomésticos          1149
Electrónicos               1772
Instrumentos musicales      753
Juguetes                   1290
Libros                      742
Muebles                    1886
Name: count, dtype: int64

### 'Fecha de Compra'

La columna "Fecha de Compra" es un caso especial porque queremos saber cuantas fechas son válidas y cuantas no, en cuyo caso las eliminariamos.

Notar que tenemos 9435 registros sin ingún valor nulo.

In [22]:
tiendas["Fecha de Compra"]

0       16/01/2021
1       18/05/2022
2       15/03/2021
3       03/05/2022
4       07/11/2020
           ...    
2353    12/02/2020
2354    17/05/2022
2355    03/05/2022
2356    15/06/2020
2357    17/11/2021
Name: Fecha de Compra, Length: 9435, dtype: object

In [44]:
fechas_convertidas = pd.to_datetime(tiendas["Fecha de Compra"], dayfirst=True)
fechas_convertidas

0      2021-01-16
1      2022-05-18
2      2021-03-15
3      2022-05-03
4      2020-11-07
          ...    
2353   2020-02-12
2354   2022-05-17
2355   2022-05-03
2356   2020-06-15
2357   2021-11-17
Name: Fecha de Compra, Length: 9435, dtype: datetime64[ns]

In [41]:
fechas_convertidas.info()

<class 'pandas.core.series.Series'>
Index: 9435 entries, 0 to 2357
Series name: Fecha de Compra
Non-Null Count  Dtype         
--------------  -----         
9435 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 147.4 KB


Notar que seguimos sin tener valores nulos después de la conversión, y que de igual forma los valores máximo y mínimo estan dentro de lo normal.

In [29]:
fechas_convertidas.describe()

count                             9435
mean     2021-08-16 07:36:48.076311808
min                2020-01-01 00:00:00
25%                2020-10-27 00:00:00
50%                2021-08-22 00:00:00
75%                2022-06-04 12:00:00
max                2023-03-31 00:00:00
Name: Fecha de Compra, dtype: object

### 'Vendedor'

In [38]:
tiendas.groupby(["Vendedor"])["Vendedor"].value_counts()

Vendedor
Beatriz Morales    650
Blanca Ramirez     679
Camila Rivera      667
Felipe Santos      701
Izabela de León    744
Juan Fernandez     694
Juliana Costa      677
Lucas Olivera      674
Maria Alfonso      679
Mariana Herrera    656
Pedro Gomez        655
Rafael Acosta      625
Santiago Silva     650
Ángel Rodriguez    684
Name: count, dtype: int64

### 'Lugar de Compra'

In [39]:
tiendas.groupby("Lugar de Compra")["Lugar de Compra"].value_counts()

Lugar de Compra
Armenia            28
Barranquilla      310
Bogotá           3943
Bucaramanga       181
Cali             1007
Cartagena         242
Cúcuta            186
Inírida            14
Leticia           142
Manizales          41
Medellín         2314
Neiva              40
Pasto              94
Pereira           517
Riohacha          163
Santa Marta       151
Soacha             19
Valledupar         14
Villavicencio      29
Name: count, dtype: int64

En mi experiencia, si hubiera valores con algún error ortográfico, tendriamos nombres que se parecerían a otros, pero con conteno más bajo, por ejemplo, "Armenia: 27" y aparte "Armenía: 1".

### 'Método de pago'

In [40]:
tiendas.groupby("Método de pago")["Método de pago"].value_counts()

Método de pago
Cupón                  504
Nequi                 1900
Tarjeta de crédito    6898
Tarjeta de débito      133
Name: count, dtype: int64

A mi consideración no hay datos extraños, por lo que no es necesario realizar un tratamiento o eliminar registros. Pero de ser el caso, probablemente hubiera usado el método de las máscaras para eliminar registros.

# Análisis de datos

## 1. Análisis de facturación



## 2. Ventas por categoría

## 3. Calificación promedio de la tienda


## 4. Productos más y menos vendidos

## 5. Envío promedio por tienda